pkg/activator/net/revision_backends.go

```diff
@@ -0,0 +1,418 @@
+
+	transport     http.RoundTripper
+	destsChan     <-chan []string
+	serviceLister corev1listers.ServiceLister
+	logger        *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev types.NamespacedName, protocol networking.ProtocolType,
+	updateCh chan<- *RevisionDestsUpdate, destsChan <-chan []string,
+	transport http.RoundTripper, serviceLister corev1listers.ServiceLister,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:           rev,
+		protocol:      protocol,
+		updateCh:      updateCh,
+		healthStates:  make(map[string]bool),
+		transport:     transport,
+		destsChan:     destsChan,
+		serviceLister: serviceLister,
+		logger:        logger,
+	}
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := make([]string, 0, len(dests))
+	for dest, healthy := range dests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+type destHealth struct {
+	dest   string
+	health bool
+}
+
+func (rw *revisionWatcher) revisionReady() bool {
+	return len(rw.dests) > 0
+}
+
+func (rw *revisionWatcher) getK8sPrivateService() (*corev1.Service, error) {
+	selector := labels.SelectorFromSet(map[string]string{
+		serving.RevisionLabelKey:  rw.rev.Name,
+		networking.ServiceTypeKey: string(networking.ServiceTypePrivate),
+	})
+	svcList, err := rw.serviceLister.Services(rw.rev.Namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(svcList) {
+	case 0:
+		return nil, fmt.Errorf("found no private service for revision %q", rw.rev.String())
+	case 1:
+		return svcList[0], nil
+	default:
+		return nil, fmt.Errorf("found multiple private services matching revision %v", rw.rev)
+	}
+}
+
+func (rw *revisionWatcher) probe(ctx context.Context, dest string) (bool, error) {
+	httpDest := url.URL{
+		Scheme: "http",
+		Host:   dest,
+	}
+	return prober.Do(ctx, rw.transport, httpDest.String(),
+		prober.WithHeader(network.ProbeHeaderName, queue.Name),
+		prober.ExpectsBody(queue.Name))
+
+}
+
+func (rw *revisionWatcher) probeClusterIP(svc *corev1.Service) (bool, string, error) {
+	if svc.Spec.ClusterIP == "" {
+		return false, "", fmt.Errorf("private service %s/%s clusterIP is nil, this should never happen", svc.ObjectMeta.Namespace, svc.ObjectMeta.Name)
+	}
+
+	svcPort, ok := GetServicePort(rw.protocol, svc)
+	if !ok {
+		return false, "", fmt.Errorf("unable to find port in service %s/%s", svc.Namespace, svc.Name)
+	}
+
+	// Context used for our probe requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	dest := net.JoinHostPort(svc.Spec.ClusterIP, strconv.Itoa(svcPort))
+	ok, err := rw.probe(ctx, dest)
+	return ok, dest, err
+}
+
+func (rw *revisionWatcher) probePodIPs() (map[string]bool, error) {
+	// Context used for our probe requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	var probeGroup errgroup.Group
+	healthStatesCh := make(chan destHealth, len(rw.dests))
```

Comment:
It's probably simpler (and cheaper) to allocate an array on the stack vs. a channel here

see: https://play.golang.org/p/Vphdv9RWee4